ocean_lib/ocean/v4/ocean_assets.py

```diff
@@ -0,0 +1,408 @@
+        if self._get_aquarius().ddo_exists(did):
+            raise AquariusError(
+                f"Asset id {did} is already registered to another asset."
+            )
+        asset.chain_id = self._web3.eth.chain_id
+        asset.metadata = metadata
+
+        asset.credentials = credentials if credentials else build_credentials_dict()
+
+        erc20_addresses = []
+        services = services or []
+        deployed_erc20_tokens = deployed_erc20_tokens or []
+        if not deployed_erc20_tokens:
+            for erc20_token_data in erc20_tokens_data:
+                erc20_addresses.append(
+                    self.deploy_datatoken(
+                        erc721_factory=erc721_factory,
+                        erc721_token=erc721_token,
+                        erc20_data=erc20_token_data,
+                        from_wallet=publisher_wallet,
+                    )
+                )
+            if not services:
+                for erc20_address in erc20_addresses:
+                    services = self._add_defaults(
+                        services, erc20_address, encrypted_files, provider_uri
+                    )
+            for erc20_token_address in erc20_addresses:
+                deployed_erc20_tokens.append(
+                    ERC20Token(self._web3, erc20_token_address)
+                )
+
+            data_tokens = self.build_data_tokens_list(
+                services=services, deployed_erc20_tokens=deployed_erc20_tokens
+            )
+        else:
+            if not services:
+                for erc20_token in deployed_erc20_tokens:
+                    services = self._add_defaults(
+                        services, erc20_token.address, encrypted_files, provider_uri
+                    )
+
+            data_tokens = self.build_data_tokens_list(
+                services=services, deployed_erc20_tokens=deployed_erc20_tokens
+            )
+
+        asset.datatokens = data_tokens
+
+        for service in services:
+            asset.add_service(service)
+
+        nft = {
+            "address": erc721_token.address,
+            "name": erc721_token.contract.caller.name(),
+            "symbol": erc721_token.symbol(),
+            "owner": publisher_wallet.address,
+            "state": 0,
+            "created": created,
+        }
+        asset.nft = nft
+
+        # Validation by Aquarius
+        validation_result, validation_errors = self.validate(asset)
+        if not validation_result:
+            msg = f"Asset has validation errors: {validation_errors}"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        # Process the DDO
+        asset_dict = asset.as_dictionary()
+        ddo_string = json.dumps(asset_dict)
+        ddo_bytes = ddo_string.encode("utf-8")
+        ddo_hash = create_checksum(ddo_string)
+
+        # Plain asset
+        if not encrypt_flag and not compress_flag:
+            encrypted_ddo = ddo_bytes
+            _ = erc721_token.set_metadata(
```

Comment:
there are multiple calls of this function that differ only by flags, data and data_hash. Do you think we can reuse the rest of the params?